python/unitytrainers/ppo/models.py

```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
+        encoded_state_list = []
+        encoded_next_state_list = []
+
+        if self.v_size > 0:
+            self.next_visual_in = []
+            visual_encoders = []
+            next_visual_encoders = []
+            for i in range(self.v_size):
+                # Create input ops for next (t+1) visual observations.
+                next_visual_input = self.create_visual_input(self.brain.camera_resolutions[i],
+                                                             name="next_visual_observation_" + str(i))
+                self.next_visual_in.append(next_visual_input)
+
+                # Create the encoder ops for current and next visual input. Not that these encoders are siamese.
+                encoded_visual = self.create_visual_observation_encoder(self.visual_in[i], 128,
+                                                                        self.swish, 1, "visual_obs_encoder", False)
+                encoded_next_visual = self.create_visual_observation_encoder(self.next_visual_in[i], 128,
+                                                                             self.swish, 1, "visual_obs_encoder", True)
+                visual_encoders.append(encoded_visual)
+                next_visual_encoders.append(encoded_next_visual)
+
+            hidden_visual = tf.concat(visual_encoders, axis=1)
+            hidden_next_visual = tf.concat(next_visual_encoders, axis=1)
+            inverse_input_list.append(hidden_visual)
+            inverse_input_list.append(hidden_next_visual)
+            encoded_state_list.append(hidden_visual)
+            encoded_next_state_list.append(hidden_next_visual)
+
+        if self.o_size > 0:
+            # Create input op for next (t+1) vector observation.
+            self.next_vector_obs = tf.placeholder(shape=[None, self.o_size], dtype=tf.float32,
+                                                  name='next_vector_observation')
+
+            # Create the encoder ops for current and next vector input. Not that these encoders are siamese.
+            encoded_vector_obs = self.create_continuous_observation_encoder(self.vector_in, 128, self.swish, 2,
+                                                                            "vector_obs_encoder", False)
+            encoded_next_vector_obs = self.create_continuous_observation_encoder(self.next_vector_obs, 128, self.swish,
+                                                                                 2, "vector_obs_encoder", True)
+
+            inverse_input_list.append(encoded_vector_obs)
+            inverse_input_list.append(encoded_next_vector_obs)
+            encoded_state_list.append(encoded_vector_obs)
+            encoded_next_state_list.append(encoded_next_vector_obs)
+
+        encoded_state = tf.concat(encoded_state_list, axis=1)
+        encoded_next_state = tf.concat(encoded_next_state_list, axis=1)
+        return encoded_state, encoded_next_state
+
+    def create_inverse_model(self, encoded_state, encoded_next_state):
```

Comment:
Good point.

Reply:
Actually, this may require a larger refactoring. I think it would be best to wait as part of a our larger changes we will be making to trainers in the future.